### PR TITLE
3417 add install agent plugin archive

### LINKED ISSUE #3417: Agent plugin upload API endpoint
## Description

Add an API endpoint that allows a user to upload an agent plugins (**content-type application/octet-stream**): `PUT /api/agent-plugins`

## Tasks
- [x] Add `store_agent_plugin(self, host_operating_system: OperatingSystem, agent_plugin: AgentPlugin)` to `IAgentPluginRepository` (0d) @ilija-lazoroski 
   - [x] Remove anything from the vulture ignore list relating to the plugin parser
   - [x] Add a stub implementation in `FileAgentPluginRepository`
- [x] Add `remove_agent_plugin(self, host_operating_system: Optional[OperatingSystem], agent_plugin_name: str, agent_plugin_type: AgentPluginType)` (0d) @ilija-lazoroski 
  - [x] This should be idempotent
  - [x] If host_operating_system is None, delete plugin for all operating systems (add to docstring)
  - [x] Add a stub implementation in `FileAgentPluginRepository`
- [x] Add `install_agent_plugin_archive(self, agent_plugin_archive: bytes)` to `IAgentPluginService` and implement it in `AgentPluginService` (0d) @ilija-lazoroski 
   - [x] Unpack/parse the archive into an AgentPlugin (plugin_archive_parser.parse_plugin)
   - [x] Call IAgentPluginRepository.remove_agent_plugin() to remove the plugin for all operating systems
   - [x] Call IAgentPluginRepository.store_agent_plugin() for each host operating system supported by the plugin
   - [x] Make this method thread-safe
- [ ] Add a new endpoint with `PUT /api/agent-plugins` (0.25d) @ilija-lazoroski

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do?

Fixes #3417 .

Adds `AgentPluginInstallation` endpoint and unit test it.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [ ] Was the documentation framework updated to reflect the changes?
* [ ] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Do all unit tests pass?
* [ ] Do all end-to-end tests pass?
* [x] Any other testing performed?
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [x] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

File: monkey/monkey_island/cc/services/agent_plugin_service/errors.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class PluginInstallError(ValueError):
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/services/agent_plugin_service/errors.py:1]: This shows great attention to detail. Good job.

File: monkey/monkey_island/cc/services/agent_plugin_service/flask_resources/agent_plugin_installation.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+
+class AgentPluginInstallation(AbstractResource):
+    urls = ["/api/agent-plugins"]
+
+    def __init__(self, agent_plugin_service: IAgentPluginService):
+        self._agent_plugin_service = agent_plugin_service
+
+    @auth_token_required
+    @roles_accepted(AccountRole.ISLAND_INTERFACE.name)
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/services/agent_plugin_service/flask_resources/agent_plugin_installation.py:23]: There's one ore more ETE tests that verify only the correct roles can access certain endpoints. We should update those tests to include this endpoint if possilbe.

File: monkey/monkey_island/cc/services/agent_plugin_service/i_agent_plugin_service.py [hunk 3]
```diff
@@ -54,5 +54,6 @@ def install_agent_plugin_archive(self, agent_plugin_archive: bytes):
         :raises RemovalError: If an error occus while attempting to uninstall a previous
         version of the plugin
         :raises StorageError: If an error occurs while attempting to install the plugin
+        :raises PluginInstallError: If an error occurs while attempting to parse the plugin archive
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/services/agent_plugin_service/i_agent_plugin_service.py:57]: Callers don't need to know this much technical detail. The error itself can contain more specifics. Plus, we may decide to raise this error for new reasons, which would make this docstring inaccurate.

```suggestion
        :raises PluginInstallError: If an error occurs while attempting to install the plugin archive
```

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/monkey_island/cc/services/agent_plugin_service/errors.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
...
+    """
+    Raised when a service ecounters an error while attempting to parse the plugin archive.
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/services/agent_plugin_service/errors.py:3]:
```suggestion
    Raised when a service encounters an error while attempting to parse the plugin archive
```

File: monkey/monkey_island/cc/services/agent_plugin_service/flask_resources/agent_plugin_installation.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+from monkey_island.cc.flask_utils import AbstractResource
+from monkey_island.cc.services.authentication_service import AccountRole
+
+from .. import IAgentPluginService
+from ..errors import PluginInstallError
+
+logger = logging.getLogger(__name__)
+
+
+class AgentPluginInstallation(AbstractResource):
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/services/agent_plugin_service/flask_resources/agent_plugin_installation.py:16]: I'm not a fan of this. This is neither RESTful nor RPC-style. Why aren't we just implementing PUT in the `AgentPlugins` resource that we already have?

File: monkey/monkey_island/cc/services/agent_plugin_service/flask_resources/agent_plugin_installation.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    urls = ["/api/agent-plugins"]
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/services/agent_plugin_service/flask_resources/agent_plugin_installation.py:17]: This URL already exists in the `AgentPlugins` resource, further supporting my point above.

File: monkey/monkey_island/cc/services/agent_plugin_service/flask_resources/agent_plugin_installation.py (hunk 5 above)
Comment (shreyamalviya) [reply monkey/monkey_island/cc/services/agent_plugin_service/flask_resources/agent_plugin_installation.py:16]: If there's a good reason for not doing that, we should change this to RPC-style (`InstallAgentPlugin`).

**Review by shreyamalviya — CHANGES_REQUESTED** [review]

File: monkey/monkey_island/cc/services/agent_plugin_service/flask_resources/agent_plugin_installation.py (hunk 6 above)
Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/services/agent_plugin_service/flask_resources/agent_plugin_installation.py:17]: That url needs expects host_os, type and plugin name. If we did the PUT there and at thiss url would not that mean that we can PUT with those arguments and GET without those. I am not really into thr REST but that is what is confusing to me.

File: monkey/monkey_island/cc/services/agent_plugin_service/errors.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mssalvatore) [reply monkey/monkey_island/cc/services/agent_plugin_service/errors.py]: +1

File: monkey/monkey_island/cc/services/agent_plugin_service/flask_resources/agent_plugin_installation.py (hunk 6 above)
Comment (mssalvatore) [reply monkey/monkey_island/cc/services/agent_plugin_service/flask_resources/agent_plugin_installation.py:17]: Ilija is correct about why we did this. That being said, making it RPC does alleviate some issues. Making it an RPC-style `/api/install-agent-plugin` is a good idea.

**Review by shreyamalviya — APPROVED** [review]

**Review by mssalvatore — APPROVED** [review]

**Pull request merged**